### PR TITLE
Stop slider animation on display and dimension changes

### LINKED ISSUE #3998: Play button sticks if all layers are deleted while playing
## 🐛 Bug

Play/Stop stops working if layer is deleted while playing

## To Reproduce

1. add stack of images as an image layer
2. hit play, then while playing delete the layer
3. add stack of images again

## Expected behavior

Play/Stop button should function normally after removing an image layer while playing

## Environment
napari: 0.4.13
Platform: Linux-3.10.0-1160.45.1.el7.x86_64-x86_64-with-glibc2.17
Python: 3.9.6 | packaged by conda-forge | (default, Jul 11 2021, 03:39:48) [GCC 9.3.0]
Qt: 5.12.9
PyQt5: 5.12.3
NumPy: 1.20.3
SciPy: 1.7.1
Dask: 2021.08.1
VisPy: 0.9.4

OpenGL:

GL version: 2.1 Mesa 18.3.4
MAX_TEXTURE_SIZE: 8192
Screens:

screen 1: resolution 1440x847, scale 1.0
Plugins:

animation: 0.0.2
cellpose-napari: 0.1.3
console: 0.0.4
napari-nikon-nd2: 0.1.3
scikit-image: 0.4.13
svg: 0.1.5

### PR DESCRIPTION
**Pull request by andy-sweet**
# Description
The original bug description covered the case when a single 3D image layer is played, removed, and then readded - in which case it was still playing.

This PR fixes that issue by stopping the playback animation whenever the number of sliders changes. When a 3D image is playing and is then removed `Dims.ndim` goes from 3 to 2.

I also handled the similar case when the display goes from 2 to 3, in which case we should also stop now.

It's possible there are some more cases I haven't thought of, or there might be a good general place to put `stop` that would catch everything, but I couldn't think of one. Open to ideas and other test cases!

## Type of change
- [x] Bug-fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# References
Closes #3998.

# How has this been tested?
- [x] added new tests to reproduce the original bug and a variant
- [ ] all tests pass with my change
- [x] I check if my changes works with both PySide and PyQt backends
      as there are small differences between the two Qt bindings.  

## Final checklist:
- [ ] My PR is the minimum possible work for the desired functionality
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] If I included new strings, I have used `trans.` to make them localizable.
      For more information see our [translations guide](https://napari.org/developers/translations.html).

### CODE REVIEW

**Comment by andy-sweet** [comment]
@tlambert03 : I assigned you to review only because I saw your name from git blame around my changes - feel free to unassign if you're unfamiliar with this code.

**Comment by tlambert03** [comment]
oh yeah... this pile of garbage is mine :joy:

File: napari/_qt/widgets/qt_dims.py [hunk 1]
```diff
@@ -97,12 +97,8 @@ def _update_range(self):
         self._resize_slice_labels()
 
     def _update_display(self):
-        """
-        Updates display for all sliders.
-
-        The event parameter is there just to allow easy connection to signals,
```
Comment (andy-sweet) [inline napari/_qt/widgets/qt_dims.py:103]: This was removed previously, so I shortened the docstring comment.

File: napari/_qt/widgets/qt_dims.py [hunk 2]
```diff
@@ -121,12 +117,8 @@ def _update_display(self):
         self._resize_slice_labels()
 
     def _update_nsliders(self):
-        """
-        Updates the number of sliders based on the number of dimensions.
-
-        The event parameter is there just to allow easy connection to signals,
```
Comment (andy-sweet) [inline napari/_qt/widgets/qt_dims.py:127]: This was removed previously, so I shortened the docstring comment.

**Review by tlambert03 — APPROVED** [review]
yep, this looks totally reasonable to me

**Comment by andy-sweet** [comment]
The test failures ([ubuntu-latest 3.9 pyqt5](https://github.com/napari/napari/runs/7102608559?check_suite_focus=true#step:8:268) and [macos-latest 3.9 pyqt5](https://github.com/napari/napari/runs/7102608810?check_suite_focus=true#step:8:260)) are new and inexplicable (at least to me...) - any ideas?

**Comment by andy-sweet** [comment]
> The test failures ([ubuntu-latest 3.9 pyqt5](https://github.com/napari/napari/runs/7102608559?check_suite_focus=true#step:8:268) and [macos-latest 3.9 pyqt5](https://github.com/napari/napari/runs/7102608810?check_suite_focus=true#step:8:260)) are new and inexplicable (at least to me...) - any ideas?

I changed the animation worker thread to connect from the finished signal to the `play_stopped` signal rather than the `play_stopped.emit` method. I think that's a potential fix because I believe Qt does a few more checks when connecting signals->slots or signals->signals rather than signals->methods/functions, though I would have expected a more severe error. But I don't know enough about the details about Qt, qtpy, pyqt, and pyside to be sure. Any insight would be appreciated.

File: napari/_qt/widgets/qt_dims_slider.py [hunk 3]
```diff
@@ -413,7 +413,7 @@ def _play(
             _connect={'frame_requested': self.qt_dims._set_frame},
         )
         worker.finished.connect(self.qt_dims.stop)
-        thread.finished.connect(self.play_stopped.emit)
+        thread.finished.connect(self.play_stopped)
```
Comment (andy-sweet) [inline napari/_qt/widgets/qt_dims_slider.py:416]: @tlambert03 : can you provide any insight on why this change might have fixed those interesting test failures.

**Comment by andy-sweet** [comment]
I marked this as a draft because I don't exactly understand my fix yet and don't want to add a flaky test to napari. When I do understand it, I'll unmark it as a draft and will likely merge it soon after.

**Comment by andy-sweet** [comment]
After looking at this fresh, I am happy enough with this.

I understand a little more about what's going on under the hood of pyqt5, when we connect to a Python callable instead of a signal. In particular, when [the callable is recognizable as a signal, the receiver is the QObject that owns that signal](https://github.com/baoboa/pyqt5/blob/11d5f43bc6f213d9d60272f3954a0048569cfc7c/qpy/QtCore/qpycore_pyqtboundsignal.cpp#L466).

When we previously connected to `play_stopped.emit`, the instance that `emit` is bound to is of the type `pyqtSignal`, which isn't recognizable as a `QObject`, at which point pyqt has to [create a slot proxy](https://github.com/baoboa/pyqt5/blob/11d5f43bc6f213d9d60272f3954a0048569cfc7c/qpy/QtCore/qpycore_pyqtboundsignal.cpp#L505) and moves it to the thread of the receiver.

My understanding ends when trying work our what `receiver` will be our case. Is it just the instance of `pyqtSignal`? In that case maybe we truly confused pyqt's guessing logic and made it behave really strangely? Note that pyside has a notion of [a global receiver](https://github.com/raccoonmonk/pyside2/blob/06e4bc4b7480cc0d9f929f34e4b34f90d7862016/sources/pyside2/PySide2/QtCore/glue/qobject_connect.cpp#L52), which may explain why it avoided similar issues.

**Comment by andy-sweet** [comment]
@tlambert03 : as you approved this already and nothing really changed, I'll likely merge this in 24 hours.

Lessons learned from this PR. Avoid connecting to plain Python callables when possible. And really avoid connecting to callables bound to instances of `pyqtSignal`!

**Comment by tlambert03** [comment]
> Lessons learned from this PR. Avoid connecting to plain Python callables when possible. And really avoid connecting to callables bound to instances of pyqtSignal!

works for me!  given the number of actively burning fires, anything that works is 100% welcome!  Merge any time!

**Pull request merged**